prog/hints.go

```diff
@@ -84,9 +84,86 @@ func generateHints(p *Prog, compMap CompMap, c *Call, arg Arg, exec func(newP *P
 }
 
 func checkConstArg(arg *ConstArg, compMap CompMap, cb func(newArg Arg)) {
-	for v, _ := range compMap[arg.Val] {
-		cb(constArg(arg.typ, v))
+	replacers := make(uint64Set)
+	for _, mutation := range mutations {
+		mutation(arg.Val, compMap, replacers)
+	}
+	for replacer, _ := range replacers {
+		// Filter out "special" values. The fuzzer will try them anyways.
+		if ok, _ := specialIntsSet[replacer]; !ok {
+			cb(constArg(arg.typ, replacer))
+		}
+	}
+}
+
+// Shrink and expand mutations model the cases when the syscall arguments
+// are casted to narrower (and wider) integer types.
+// ======================================================================
+// Motivation for shrink:
+// void f(u16 x) {
+//		u8 y = (u8)x;
+//		if (y == 0xab) {...}
+// }
+// If we call f(0x1234), then we'll see a comparison 0x34 vs 0xab and we'll
+// be unable to match the argument 0x1234 with any of the comparison operands.
+// Thus we shrink 0x1234 to 0x34 and try to match 0x34.
+// If there's a match for the shrank value, then we replace the corresponding
+// bytes of the input (in the given example we'll get 0x12ab).
+// Sometimes the other comparison operand will be wider than the shrank value
+// (in the example above consider comparison if (y == 0xdeadbeef) {...}).
+// In this case we ignore such comparison because we couldn't come up with
+// any valid code example that does similar things. To avoid such comparisons
+// we check the sizes with leastSize().
+// ======================================================================
+// Motivation for expand:
+// void f(i8 x) {
+//		i16 y = (i16)x;
+//		if (y == -2) {...}
+// }
+// Suppose we call f(-1), then we'll see a comparison 0xffff vs 0xfffe and be
+// unable to match input vs any operands. Thus we sign extend the input and
+// check the extension.
+// As with shrink we ignore cases when the other operand is wider.
+// Note that executor sign extends all the comparison operands to int64.
+// ======================================================================
+func shrinkExpand(v uint64, compMap CompMap, replacers uint64Set) {
+	// Map: key is shrank/extended value, value is the maximal number of bits
+	// that can be replaced.
+	res := make(map[uint64]uint)
+	for _, size := range []uint{8, 16, 32} {
+		res[v&((1<<size)-1)] = size
+		if v&(1<<(size-1)) != 0 {
+			res[v|^((1<<size)-1)] = size
+		}
+	}
+	res[v] = 64
+
+	for mutant, size := range res {
+		for newV, _ := range compMap[mutant] {
+			if leastSize(newV) <= size {
```

Comment:
This can be implemented in 1 line (assuming you declare mask above this):
```
if newHi := newV & ^mask; newHi == 0 || newHi^^mask == 0 {
```
This will get rid of leastSize.